Implementation of discussed changes

vidartf opened the pull request:
This refactors widgets, and should implement most of the discussed changes of the ROI. Some needed cross-contamination with NEW_Events and NEW_Interactive.

vidartf committed:
Merge branch 'NEW_Events' into NEW_ROI

vidartf committed:
Updated widget code to use events + refactored

vidartf committed:
More widget code + bugfixes

Fixes misplaced bottom left resizers, and now allows setting coordinates
before being added to a plot. Also added more interface functions for
setting values (set_bounds).

vidartf committed:
Added 'out' keyword awareness for Interactive

Added possiblity to use interactive with existing signal by it picking
up on the 'out' keyword being passed to the function. No such
functionality for positional args naturally.

vidartf committed:
Bottom left origin

Switched ROI origin to be bottom left, as is the matplotlib convention.

vidartf committed:
Added bounds check flag

If changing left,right from 4,7 to 10,13, then previously you would have
to set right first, or left would just keep its old value (as it was
greater than right). If going the other way, then left would need to be
set first. To avoid having to check this when setting both, an internal
flag is added, so that the bounds cheking can be disabled. Note: Use the
flag responsibly!

vidartf committed:
add_widget function

Added an add_widget function, which will add a widget representing the
ROI to a signal's plot. Which plot/axes the ROI maps to are specified by
the axes argument, which defaults to: the first two nav axes if nav dim
>= 2; the first two sig axes if sig dim >= 2 (else raises excpetion).

vidartf committed:
Interactive util function

Added utility function on ROI to get an interactive slice of passed
signal.

vidartf committed:
Added "navigation" to interactive

Made ROI.interactive add a widget to 'navigation_signal', if supplied.
If 'navigation_signal' is "same", then 'navigation_signal' will use the
value of 'signal' argument.

vidartf committed:
Removed redundant 'obj' from interactive

Argument 'obj' was not used for anything...

vidartf committed:
Changed HE signal_data_function to lazy

Previously, the signal_data_function in HyperExplorer was only used in
plot(), where it was copied to the SpectrumLine/Image. Changing it
afterwards had no effect. With this change, the spectrumLine/Image will
always use the signal_data_function on HyperExplorer. This could maybe
be achieved with lambda, but I felt def was clearer. Should maybe be
done for navigation_data_function as well.

vidartf committed:
Changed back to top<bottom + widget bounds checks fixes

vidartf committed:
Added ROI 'axes' parameter

Added 'axes' parameter to ROI.__call__, so it knows which axes to slice.

vidartf committed:
Added ROI.add_widget color parameter.

vidartf committed:
Added ROI.navigate utility

vidartf committed:
Event suppression safety feature

Added a boolean attribute to make sure an EventSuppressionContext only
does it __exit__ cleanup once.

vidartf committed:
Unused attribute?

vidartf committed:
Refactored widgets

Refactored widget, with special focus on extracting common features up
the class hierarchy. Also moved navigation functionality into widgets
again, but made it so it has to be enabled explicitly
(connect_navigation). Now, ROI navigation only have to care about it's
connection to the widget.

vidartf committed:
Small bugfixes

vidartf committed:
Allow Interactive.update to receive args (ignored)

If Interactive.update can receive and ignore args, it can also be
connected to events that have an argument. Alternatively, we could pass
these to Interactive.f, or maybe a better long term solution: Implement
somthing similar to pyqt signals: event[bool].connect() to connect a
function that expects a boolean argument, event.connect() to connect a
function that expects no args. Alternatively, we could use numbers to
specify number of arguments, but type seems more powerful (although
type-checking isn't very pythonic).

vidartf committed:
_apply_funtion_on_... handles shape change

If the shape of 'out' signal changes as a consequence of
Signal._apply_function_on_data_and_remove_axis, then it now calls
out.get_dimensions_from_data() to update the axes_manager.
get_dimensions_from_data is also decorated with @auto_replot, so if
previously plotted, it will trigger a replot.

vidartf committed:
Includes all return paths

Fixes missing function return path in
06ff12c0ea3f6f0a28402bc1306fabbc46201a8e

vidartf committed:
Added out argument to ROI.interactive

vidartf committed:
Revert "Changed HE signal_data_function to lazy"

This reverts commit d54b8a770589619002bf49751defce126e2894b3.

vidartf committed:
Moved ROI navigation to separate branch

see vidartf/hyperspy/NEW_ROI_navigation

vidartf committed:
Refactored widgets

Refactored a lot of widget code, with focus on reusability for different
kinds of widgets.

francisco-dlp commented:
This is awesome, I thought that it'll take us weeks to get all of this done and you've done it in days (!). I bet that you have worked in multiple branches and merged them all into NEW_ROI, could you pull request to the individual branches too? In this way we'll keep things tidy what should make it easier for others to  discuss and review your work.

vidartf commented:
I mostly worked in one branch, but I'll cherry-pick the commits out to the
corresponding branches and PR those.

vidartf committed:
1D span widget + small fixes

Made a 1D span widget base on ModifiableSpanSelector, but which
implements the ResizableDraggablePatchBase interface. Also did some
small fixes in widget code.

vidartf committed:
Added 1D SpanROI

Implemented SpanROI for 1D ROI selection. Made a lot of the code common
for ND, and moved it into BaseROI. We might have to segregate BaseROI
and common ancestor for 1D/2D later on (maybe for circle ROI?), but for
now this is OK.

Conflicts:
	hyperspy/roi.py

vidartf committed:
Bugfix ROI

Slipped through the cracks when cherry-picking.

vidartf commented:
Also added 1D ROI with appropriate widget.

vidartf committed:
Added 1D ROI stepper

Make 1D ROI widget snap to data steps.

This will have to be modified if non-linear axes are introduced!

francisco-dlp committed:
Remove _suspend and _resume keeping the functionality.

francisco-dlp commented on hyperspy/drawing/widgets.py:
```diff
@@ -67,40 +71,48 @@ def set_on(self, value):
         self.patch.set_animated(hasattr(ax, 'hspy_fig'))
 
-    def add_axes(self, ax):
+    def set_axis(self, ax):
```
I agree that `add_axes` was misleading. However, `ax` should be a MPL Axes instance, so `set_axis` doesn't feel fully right either. What about `set_mpl_axes` or `set_mpl_ax`?

francisco-dlp commented:
Nice.

As previously discussed, this branch has a massive amount of changes and it would have been better to perform many of these changes in separate branches. For example, here, in addition to implementing the ROIs, you have fixed bugs, renamed attributes, enhanced some features and added others... All in all, it is an amazing contribution. However, it is difficult to review and to discuss the changes when they are all coming in one PR. If you had made several PRs many of them would be already merged, however, as it is, we have to review and agree on the full set. Beside, renaming attributes is often dangerous, specially in the drawing code that does not yet benefit from unittests. So, renaming existing attributes should be done in isolated branches and we should always make sure that we perform the renaming in all files.

vidartf committed:
Merge pull request #4 from francisco-dlp/NEW_ROI_w/t_suspend

Remove _suspend and _resume keeping the functionality.

vidartf committed:
Removed unneeded Events suppression.

vidartf commented:
Agree on most of your points, I'll just have to figure out the hierarchy of which changes that depend on each other. Do you think I should push here, or to hyperspy master?

francisco-dlp commented:
It depends, the commits that deal only the new ROIs should come here, but everything else should go to hyperspy/master.

Sorry that we don't have a good Guide for developers yet, I just opened #445 to fix this.

vidartf commented:
Well, I tried to move some stuff out, but most of what happens in widget.py and roi.py is interdependent, especially with the big revamp of widget.py that was needed to rewire everything. I'm not sure I can break it down into meaningful chunks that won't have a broken intermediate state.

vidartf commented:
As an example, in 382de116e36de4a91b6e0b5cf47a7e707938cae4 I moved _a lot _ of code around to extract common tasks higher in widget class hierarchy, which of source makes the diff a mess. And the state of the code before that was rather WIP. If all those changes have to broken down into separate items, each with a workable code base, then I simply won't have the time necessary. To be honest, I have lost oversight over all the different pull requests, and which branches have which commits etc. Hopefully that doesn't mean any bugs will slip through, but the odds certainly have gone up.

On the other hand, if I could help code review by adding docstrings then that's something I'd be happy to do.

francisco-dlp commented:
The situation has vastly improved, thanks!

francisco-dlp commented:
Could you merge master into your "NEW_ROI" branch? It is not straightforward and you're in a better position than I to perform the task without breaking something...

vidartf committed:
Bugfixes new Widgets

Several smaller bugs fixed in new widget implementations. Most were
innocent, but a few could lead to the widgets just throwing exceptions
when interacting with them.

vidartf committed:
NEW: ModifiableSpanSelector.can_switch

Added an attribute and accompanying methods on ModifiableSpanSelector to
allow left and right edge modification to switch if the mouse tries to
cross the edge. The left value will still always be the smallest, simply
which edge the mouse is interacting with will change. the default value
for the attribute is False, which will correspond to the old behavior.
However, DraggableResizableRange will set it to True for it's use, to
correspond to the similar behavor for other widgets in the same
hierarchy.

vidartf committed:
BUG: Fixed ModifiableSpanSelector.tolerance use

Previously, ModifiableSpanSelector.tolerance was not applied as one
would expect for use of contains() and picking (although picking was not
really in use). This commit fixes these issues by overriding the
contains() method on it's Rectangle instance.

vidartf committed:
ENH/BUG: Renamed add_axes/set_axis to set_mpl_axis

Cleaned up rename code.

vidartf commented:
That should do it, right?

vidartf commented:
No, I still need to fix some stuff, let me do it quickly.

vidartf committed:
Merge branch 'master' into NEW_ROI

Conflicts:
	hyperspy/drawing/widgets.py

vidartf committed:
Merge remote-tracking branch 'francisco-dlp/NEW_ROI' into NEW_ROI

vidartf commented:
There, I also merged your branch so that it can auto-merge,

vidartf committed:
BUG: Dangling rename

vidartf committed:
ENH: Whitespace cleanup

vidartf committed:
Removed contentious issue

What to do here is heavily discussed elsewhere
(francisco-dlp/hyperspy#6), so I will move it to separate branch/PR.

vidartf commented:
Now I also removed the code relevant to discussion here francisco-dlp/hyperspy#6. Probably better to have that as separate PR :)

Now I think you can merge, if you approve so far. I'll try to doc up the code as soon as I have time, but there will be a bit of a crunch for me going forward.

vidartf committed:
BUG: Trailing renames

Slipped through during merges...

vidartf committed:
Model fix

vidartf committed:
WIP

vidartf committed:
Finished off PointROIs

Also refactored bits of the internal mechanics of ROIs.

vidartf committed:
NEW: Finished implementing DraggableResizable2DLine.

Also made widget coordinates a property.

vidartf committed:
Merge branch 'master' into NEW_line_ROI

vidartf committed:
BUG: ROI fixups

vidartf committed:
BUG: Cleanup Widgets

vidartf committed:
WIP: LineROI width indicators.

vidartf committed:
BUG: Changes following new Event signatures

vidartf committed:
Merge remote-tracking branch 'francisco-dlp/NEW_Events' into NEW_ROI

Conflicts:
	hyperspy/tests/test_events.py

vidartf committed:
Cleanup tree

vidartf committed:
Merge remote-tracking branch 'francisco-dlp/NEW_ROI' into NEW_ROI

vidartf committed:
More cleanup

vidartf committed:
Cleanup

vidartf commented:
I'm trying to separate the pure ROI code and widget code now. This PR should now only be ROI code. The diff will be a lot simpler if you merge your NEW_Events first.

vidartf committed:
Cleanup by merge

Conflicts:
	hyperspy/drawing/widgets.py
	hyperspy/roi.py

vidartf commented:
For now, I've put all the widget code here:
https://github.com/vidartf/hyperspy/tree/NEW_widgets2

vidartf commented:
Finally, I've put the code that connects the new ROIs with the new widgets here:
https://github.com/vidartf/hyperspy/tree/NEW_ROI_interactive3

vidartf committed:
Merge remote-tracking branch 'upstream/master' into NEW_ROI

vidartf committed:
Cleanup: Removed remnant signal_map

vidartf committed:
ENH: Improved internal ROI code

vidartf committed:
DOC: Docstrings for ROI code

vidartf committed:
NEW: Event.suppress_single()

vidartf committed:
Merge branch 'NEW_Events_signatures' into NEW_Events_suppress_single

Conflicts:
	hyperspy/events.py

vidartf committed:
Typo

vidartf committed:
BUG: Syntax remnants

vidartf committed:
ENH: Removed kwargs from trigger

Until a reasonable standard for counting args is agreed upon, keyword arguments cannot be passed to trigger().

vidartf committed:
DOC: Document Events

vidartf committed:
Merge branch 'NEW_Events_suppress_single' into NEW_ROI

Conflicts:
	hyperspy/tests/test_events.py

vidartf committed:
ENH: mix_axes

vidartf committed:
Merge branch 'master' into NEW_ROI

vidartf committed:
Refactored ROI._parse_axes

vidartf committed:
ENH: PEP8

vidartf committed:
Cleanup

vidartf committed:
BUG: Fixed Point2DROI

vidartf committed:
BUG: Refix

vidartf committed:
BUG: Fix ROI._parse_axes

dnjohnstone commented:
For some things I'm working on at the moment I have a stack of images and I want to plot the intensity in a subset of pixels for each image as an array with dimensions equal to my navigation dimensions - I understand that this is essentially done somewhere in this PR and all I'd really like to do is extend the shapes of the subset of pixels that are being plotted e.g. a circle around a point, an annulus around the centre...BUT I don't want to make this branch any more complicated than it already is so @vidartf I was wondering if you've got plans to separate things out any further?

vidartf commented:
@dnjohnstone I'm afraid my earlier attempt at separating things is horribly dated (and quite possibly broken) due to the branch hell it led to. I still have workable "higher level" branches though, so I should be able to extract a branch for what you need. From how you describe it, and what I know of your work, you don't need any widgets to manually interact with the ROIs? If so, I can just isolate the code for the ROIs. I have a circular ROI currently, but not ellipsis yet. Let me know if you need that and I'll rush it.

Technical note: Currently there is no support for "soft" edges, i.e. the pixels/voxels are either selected or not. Adding feathering is of course possible, but I/we would have to figure out how to best implement it (partial masking / alpha channel would be needed).

vidartf commented:
I think a decent ROI-only implementation is now in [vidartf/hyperspy/NEW_ROI_only](https://github.com/vidartf/hyperspy/tree/NEW_ROI_only).

dnjohnstone commented:
Ok thanks for that! That's exactly right, I only need the ROI stuff for now so that's perfect. I'll have a look in detail soon.

Regarding soft edges: This is an interesting point for masking when doing decompositions as well. By coincidence I was just thinking of looking in to smooth edges for masks because I'm using a circle there and it's just ugly without soft edges - would probably be neat if we can settle on the best way to do this and use it generally.
